sm_ipt/sm_ipt.c

```diff
@@ -0,0 +1,450 @@
+		return;
+	}
+
+	do {
+		do {
+			/* create shifted mask with bits set where `blocks` can be allocated */
+			cur_mask = free_mask_get(ctx);
+			sh_mask = cur_mask;
+			for (size_t i = 1; i < blocks; i++) {
+				sh_mask &= (sh_mask << 1);
+			}
+
+			 /* if no memory */
+			if (sh_mask == 0) {
+				/* wait for any block to be empty */
+				sm_ipt_os_take(&ctx->out_sem);
+				sem_taken = true;
+			}
+
+		} while (sh_mask == 0);
+
+		/* get first available blocks */
+		free_index = mask_clz(sh_mask);
+		/* create bit mask with blocks that will be used */
+		mask = calc_mask(blocks, free_index);
+		/* update masks */
+		unset_success = free_mask_unset(ctx, mask);
+
+		/* there is a small probability that unset will be unsuccessful */
+		if (!unset_success) {
+			/* give semaphore, because free_mask_unset may cause other thread
+			 * waiting before it reverted the changes
+			 */
+			SM_IPT_DBG("Not enough free shared memory. Waiting.");
+			sm_ipt_os_give(&ctx->out_sem);
+			sem_taken = false;
+		}
+	} while (!unset_success);
+
+	/* Give semaphore back, because there may be some other thread waiting */
+	if (sem_taken && ((cur_mask & ~mask) != 0)) {
+		sm_ipt_os_give(&ctx->out_sem);
+	}
+
+	uint32_t *mem_start = (uint32_t *)&ctx->out.shm_queue->allocable[
+								ctx->out.block_size * free_index];
+
+	mem_start[0] = blocks * ctx->out.block_size;
+
+	*buf = (uint8_t *)(&mem_start[1]);
+	SM_IPT_DBG("Allocated memory at 0x%x", (uint32_t)*buf);
+}
+
+void sm_ipt_free_tx_buf(struct sm_ipt_ctx *ctx, uint8_t *buf)
+{
+	uint32_t *mem_start = (uint32_t *)buf - 1;
+	uint32_t offset = (uint8_t *)mem_start - (uint8_t *)ctx->out.shm_queue->allocable;
+	uint32_t block_index = offset / ctx->out.block_size;
+	uint32_t allocated_size;
+	uint32_t allocated_blocks;
+
+	SM_IPT_ASSERT(block_index < NUM_BLOCKS);
+	SM_IPT_ASSERT(buf == &ctx->out.shm_queue->allocable[block_index * ctx->out.block_size]);
+
+	allocated_size = mem_start[0];
+	allocated_blocks = allocated_size / ctx->out.block_size;
+
+	SM_IPT_ASSERT(allocated_blocks % ctx->out.block_size == 0);
+	SM_IPT_ASSERT(allocated_size <= ctx->out.allocable_size);
+	SM_IPT_ASSERT(offset + allocated_size <= ctx->out.allocable_size);
+
+	free_mask_set(ctx, calc_mask(allocated_blocks, block_index));
+	sm_ipt_os_give(&ctx->out_sem);
+	SM_IPT_DBG("Dealocated TX block at 0x%x", (uint32_t)buf);
+}
+
+
+static void queue_send(struct sm_ipt_ctx *ctx, uint8_t data)
+{
+	uint32_t tx;
+	uint32_t dst;
+
+	sm_ipt_os_lock(&ctx->out_mutex);
+
+	tx = ctx->out.shm_queue->queue_tx;
+	dst = tx;
+
+	if (dst >= QUEUE_ITEMS_CNT) {
+		memory_corrupted_error();
+		return;
+	}
+
+	tx++;
+
+	if (tx >= QUEUE_ITEMS_CNT) {
+		tx = 0;
+	}
+
+	ctx->out.shm_queue->queue[dst] = data;
+
+	SM_IPT_OS_MEMORY_BARRIER();
+
+	ctx->out.shm_queue->queue_tx = tx;
+
+	sm_ipt_os_unlock(&ctx->out_mutex);
+
+	sm_ipt_os_signal(&ctx->os_ctx);
+}
+
+
+int sm_ipt_send(struct sm_ipt_ctx *ctx, uint8_t *buf, size_t len)
+{
+	uint32_t *mem_start = (uint32_t *)buf - 1;
```

Comment:
I see a repeated pattern when you subtract 1. Is it a special value which can be defined ?